adobe/aio-cli-plugin-app-dev
# feat: ACNA-3098 - allow action invocation of ES Modules / Typescript

closes #69
closes #42

Enables ESM and TypeScript support for action code, via webpack.

## Pre-requisite PRs

- [x] https://github.com/adobe/aio-lib-runtime/pull/192

## Related PRs

- [x] https://github.com/adobe/generator-aio-app/pull/262
- [x] https://github.com/AdobeDocs/app-builder/pull/350

## TODO

- [x] add an action file watcher so files can be rebuilt (similar to `aio app run`). See [this](https://github.com/adobe/aio-cli-plugin-app/blob/master/src/lib/actions-watcher.js). This may introduce latency
- [X] ~remove the use of the magic string for the built action code in `dist`, we may have to call the lib to build the actions directly, and not use `aio app build`. See [this](https://github.com/adobe/aio-cli-plugin-app/blob/74ebd4d396dc8fb91bb193c3e74e7a4fec7e5513/src/lib/build-actions.js#L23).~ We can't do this, since aio-lib-runtime uses [this magic string](https://github.com/adobe/aio-lib-runtime/blob/da90917ef69a02b0131d3abafa123f1b503948ae/src/build-actions.js#L173) and does not expose it
- [x] use aio-lib-runtime to build directly
- [x] debugging via VS Code must work properly (support source maps)
- [x] for Typescript, you will need to npm install (devDep) `typescript, ts-loader` and have a custom `webpack-config.js` (see example project and the [docs](https://developer.adobe.com/app-builder/docs/guides/configuration/webpack-configuration/)). Any includes/excludes should be in `tsconfig.json`. `tsconfig.json` should have the option `sourceMap: true`
- [X] update AB Docs (see PR above)

## How Has This Been Tested?

1. `aio plugins install https://github.com/adobe/aio-cli-plugin-app-dev#story/ACNA-3098`
2. `git clone https://github.com/shazron/test-debugger.git` (from [this repo](https://github.com/shazron/test-debugger))
3. `cd test-debugger`
4. `aio app dev` in the cloned repo folder OR
5. step by step debug the 3 actions in Visual Studio Code via the Debug tab (put breakpoints in the CommonJS, ESM, and TypeScript actions)
6. while the command is still running, make edits to the action code, note the code gets built and updated in the log. Editing and saving while step-by-step debugging is undefined behaviour and not supported.
7. make sure that the edits are reflected in the action results after calling the actions
8. test web-src debugging
9. to uninstall this plugin: `aio plugins uninstall @adobe/aio-cli-plugin-app-dev`

## Testing in your App

1. Update your app's `.vscode/launch.json`, see: https://developer.adobe.com/app-builder/docs/guides/development/#debugging-with-vs-code
2. `aio plugins install https://github.com/adobe/aio-cli-plugin-app-dev#story/ACNA-3098`
3. Add/edit your [webpack-config.js and tsconfig.json](https://github.com/adobe/aio-cli-plugin-app-dev/blob/13cd91c80c35d6376118ec720a1711d67c5035a8/README.md) files
4. `aio app dev` in the your app root OR VS Code debug via the Debug Tab
5. to uninstall this plugin: `aio plugins uninstall @adobe/aio-cli-plugin-app-dev`


## Types of changes




## Checklist:
